src/gui/qgsexpressionpreviewwidget.cpp

```diff
@@ -49,20 +49,7 @@ void QgsExpressionPreviewWidget::setCurrentFeature( const QgsFeature &feature )
 {
   // todo: update the combo box if it has been set externaly?
 
-  // force the feature to be valid, so it can evaluate an invalid feature but having its fields set
-  if ( !feature.isValid() )
-  {
-    QgsFeature validFeature( feature );
-    validFeature.setValid( true );
```

Comment:
if you remove this, that will bring back the issue when a layer is empty and you try an expression (https://github.com/qgis/QGIS/issues/37100).
Or am I missing something?

Reply:
With this pull request, if the layer is empty (or you specify an invalid FID),
- if the expression doesn't use fields, then it is evaluated
- if the expression uses fields, the preview label displays "No feature was found on this layer to evaluate the expression" (not that the expression is invalid)

Reply:
ok, that makes sense.